### ISSUE TITLE
Spurious E0072 error in const eval: recursive type has infinite size

### ISSUE THREAD

**Issue by jswrenn** [rust-lang/rust#117518]
Compiling this code:
```rust
trait Layout {
    const ALIGN: usize;
}

struct Struct;

impl Layout for Struct {
    const ALIGN: usize = {
        #[repr(C)]
        struct SizeIsAlign {
            _byte: u8,
            _aligned: [Self; 0], // change this to `[Struct; 0]`
        }
        
        core::mem::size_of::<SizeIsAlign>()
    };
}
```
produces the spurious compiler error:
```text
error[E0072]: recursive type `SizeIsAlign` has infinite size
  --> src/lib.rs:10:9
   |
10 |         struct SizeIsAlign {
   |         ^^^^^^^^^^^^^^^^^^
11 |             _byte: u8,
12 |             _aligned: [Self; 0], // change this to `[Struct; 0]`
   |                       --------- recursive without indirection
   |
help: insert some indirection (e.g., a `Box`, `Rc`, or `&`) to break the cycle
   |
12 |             _aligned: Box<[Self; 0]>, // change this to `[Struct; 0]`
   |                       ++++         +
```
This error is spurious because `SizeIsAlign` does *not* have infinite size. 

A simple change — replacing `Self` with `Struct` bypasses the error:
```rust
impl Layout for Struct {
    const ALIGN: usize = {
        #[repr(C)]
        struct SizeIsAlign {
            _byte: u8,
            _aligned: [Struct; 0], // change this to `[Struct; 0]`
        }
        
        core::mem::size_of::<SizeIsAlign>()
    };
}
```
Or, alternatively, adding a little indirection also works:
```rust
impl Layout for Struct {
    const ALIGN: usize = {
        #[repr(C)]
        struct SizeIsAlign<T> {
            _byte: u8,
            _aligned: [T; 0],
        }
        
        core::mem::size_of::<SizeIsAlign<Self>>()
    };
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.75.0-nightly (608e9682f 2023-10-29)
binary: rustc
commit-hash: 608e9682f0a6482903de8d3332770104a0ad943c
commit-date: 2023-10-29
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.3
```

**Comment by matthewjasper** [comment]
`Self` resolves to `SizeIsAlign` here, and the error stays when making that change

**Comment by jswrenn** [comment]
D'oh.
